python/lsst/ctrl/mpexec/mpGraphExecutor.py

```diff
@@ -157,57 +162,72 @@ class _JobList:
-
-        Returns
-        -------
-        QuantumNodes : `set` [`~lsst.pipe.base.QuantumNode`]
-            Set of QuantumNodes that have successfully finished
-        """
-        return set(job.qnode for job in self.jobs if job.state == JobState.FINISHED)
-
-    def failedNodes(self):
-        """Return set of jobs IDs that failed for any reason.
+    def submit(self, job, butler, quantumExecutor, startMethod=None):
+        """Submit one more job for execution
 
-        Returns
-        -------
-        QuantumNodes : `set` [`~lsst.pipe.base.QuantumNode`]
-            Set of QUantumNodes that failed during processing
+        Parameters
+        ----------
+        job : `_Job`
+            Job to submit.
+        butler : `lsst.daf.butler.Butler`
+            Data butler instance.
+        quantumExecutor : `QuantumExecutor`
+            Executor for single quantum.
+        startMethod : `str`, optional
+            Start method from `multiprocessing` module.
         """
-        return set(job.qnode for job in self.jobs
-                   if job.state in (JobState.FAILED, JobState.FAILED_DEP, JobState.TIMED_OUT))
+        # this will raise if job is not in pending list
+        self.pending.remove(job)
+        job.start(butler, quantumExecutor, startMethod)
+        self.running.append(job)
 
-    def timedOutIds(self):
-        """Return set of jobs IDs that timed out.
+    def setJobState(self, job, state):
+        """Update job state.
 
-        Returns
-        -------
-        jobsIds : `set` [`int`]
-            Set of integer job IDs.
+        Parameters
+        ----------
+        job : `_Job`
+            Job to submit.
+        state : `JobState`
+            New job state, note that only FINISHED, FAILED, TIMED_OUT, or
+            FAILED_DEP state is acceptable.
         """
-        return set(job.qnode for job in self.jobs if job.state == JobState.TIMED_OUT)
+        allowedStates = (
+            JobState.FINISHED,
+            JobState.FAILED,
+            JobState.TIMED_OUT,
+            JobState.FAILED_DEP
+        )
+        assert state in allowedStates, f"State {state} not allowed here"
+
+        # remove job from pending/running lists
+        if job.state == JobState.PENDING:
+            self.pending.remove(job)
+        elif job.state == JobState.RUNNING:
+            self.running.remove(job)
+
+        qnode = job.qnode
+        # it should not be in any of these, but just in case
+        self.finishedNodes.discard(qnode)
+        self.failedNodes.discard(qnode)
+        self.timedOutNodes.discard(qnode)
+
+        job._state = state
+        if state == JobState.FINISHED:
+            self.finishedNodes.add(qnode)
+        elif state == JobState.FAILED:
+            self.failedNodes.add(qnode)
+        elif state == JobState.FAILED_DEP:
+            self.failedNodes.add(qnode)
+        elif state == JobState.TIMED_OUT:
+            self.failedNodes.add(qnode)
+            self.timedOutNodes.add(qnode)
```

Comment:
can you put an else at the end of this that raises? I know that it shouldn't get to it if the assert works above but in the future someone might clean something up that removes it.